### PR TITLE
Add protocol support for WebTransport.

### PR DESCRIPTION
**Pull request by jianjunz**

### CODE REVIEW

File: doc/Client-Portal Protocol.md [hunk 1]
```diff
@@ -164,7 +166,8 @@ This a format for client reconnects.
           {
            id: string(StreamID),
            type: "forward" | "augmented" |"mixed",
-           media: object(MediaInfo),
+           media: object(MediaInfo) | null,
+           data: object(DataInfo) | null,
```
Comment (taste1981) [inline doc/Client-Portal Protocol.md]: Please add define of DataInfo

Comment (jianjunz) [reply doc/Client-Portal Protocol.md]: Changed to boolean since `Publication.data` is boolean. Do you think it should be an object to describe the stream? I though a stream is always ordered and reliable. So it sounds like no more description is needed.

Comment (taste1981) [reply doc/Client-Portal Protocol.md]: Yes i agree it can be boolean only. How about move the data property to MediaInfo?

Comment (jianjunz) [reply doc/Client-Portal Protocol.md]: Data stream could carry packets other than media frames, so I think it's not a part of media information.

File: doc/Client-Portal Protocol.md [hunk 2]
```diff
@@ -350,38 +356,44 @@ This a format for client reconnects.
...
 **RequestData**: The PublicationRequest object with following definition:
 
+```
   object(PublicationRequest)::
     {
-     media: object(WebRTCMediaOptions),
-     attributes: object(ClientDefinedAttributes)
+     media: object(WebRTCMediaOptions) | null,
+     data: true | false,
+     transport: object(TransportOptions),
```
Comment (starwarfan) [inline doc/Client-Portal Protocol.md]: Previous publications can be "webrtc", "streaming" or "sip".
Are the "data" and "transport" fields for "webrtc" or "quic-p2p" only?
I'm not sure, is a quic-connection sending media data a "quic-p2p" publication or "webrtc" publication?

File: doc/Client-Portal Protocol.md [hunk 3]
```diff
@@ -350,38 +356,44 @@ This a format for client reconnects.
...
+            | false
+    }
+```
+
 **ResponseData**: The PublicationResult object with following definition if **ResponseStatus** is “ok”:
 
   object(PublicationResult)::
     {
-     id: string(SessionId) //will be used as the stream id when it gets ready.
+      transportId: string(transportId),  // Can be reused in the following publication or subscription.
```
Comment (starwarfan) [inline doc/Client-Portal Protocol.md]: I think this change is also for addTransceiver on single connection, how about put this in another PR?

File: doc/Client-Portal Protocol.md [hunk 4]
```diff
@@ -542,24 +556,124 @@ This a format for client reconnects.
...
+      status: "soac" | "ready" | "error",
+      data: object(OfferAnswer) | object(CandidateMessage) | object(P2PQuicParametersMessage) /*If status equals “soac”*/
+          | (undefined/*If status equals “ready” and session is NOT for recording*/
+          | string(Reason)/*If status equals “error”*/
+    }
 
   object(SessionProgress)::
     {
      id: string(SessionId), /* StreamId returned in publishing or SubscriptionId returned in subscribing*/
      status: "soac" | "ready" | "error",
```
Comment (starwarfan) [inline doc/Client-Portal Protocol.md]: Is "soac" in "SessionProgress" still needed If the "TransportProgress" has the "soac" (offer, answer, candidate),?

File: doc/Client-Portal Protocol.md [hunk 5]
```diff
@@ -164,7 +169,8 @@ This a format for client reconnects.
           {
            id: string(StreamID),
            type: "forward" | "augmented" |"mixed",
-           media: object(MediaInfo),
+           media: object(MediaInfo) | null,
+           data: true | false,
```
[the remark this answers is not in the document]
Comment (jianjunz) [reply doc/Client-Portal Protocol.md]: I think current it's either a media stream or a data stream. Or a media stream is a special data stream since the data is media frames?

File: doc/Client-Portal Protocol.md (hunk 2 above)
Comment (jianjunz) [reply doc/Client-Portal Protocol.md]: The idea of adding a `transport` is to make it possible to reuse transport channels. I'm not sure if SIP or other streaming transport channels could be reused.

Sending media data through QUIC connection is a "quic-p2p" publication, although it's still belongs to WebRTC right now. As we decided to create a new agent for QUIC instead of put it in WebRTC agent, we need a new type of QUIC connections.

File: doc/Client-Portal Protocol.md (hunk 3 above)
Comment (jianjunz) [reply doc/Client-Portal Protocol.md]: Adding `transportId` may not enough for reuse PeerConnection for multiple RTP channels. This PR only adds it for QUIC connections. You may need more changes to make it possible for combining PeerConnections. And `addTransceiver` at client side doesn't reuse PeerConnection in early stage.

File: doc/Client-Portal Protocol.md (hunk 4 above)
Comment (jianjunz) [reply doc/Client-Portal Protocol.md]: It sounds like recorder info is also soac? Should we rename it to another name?

File: doc/Client-Portal Protocol.md (hunk 2 above)
Comment (starwarfan) [reply doc/Client-Portal Protocol.md]: I mean the description here is not clear.
Can "webrtc" transport publications have "data" field? The "data" is only needed by newly added "quic-p2p" publications?
Can these fields be optional? In my understanding, the "data" and "media" fields are exclusive, one publication can not have them both. Is a publication allowed with "media" null and "data" false? And a "webrtc" transport publication's "media" can not be null? I think you'd better add more comment in document or adjust the format to make it more clear.

File: doc/Client-Portal Protocol.md [hunk 6]
```diff
@@ -542,24 +556,124 @@ This a format for client reconnects.
        sdpMLineIndex: number(mLineIndex),   // optional in RemovedCandidatesMessage
        candidate: string(candidateSdp)
       }
+
+    object(TransportOptions)::
+      {
+        type: "webrtc" | "quic-p2p",
+        id: string(transportId) | null,  // null will result to create a new transport channel.
```
Comment (starwarfan) [inline doc/Client-Portal Protocol.md]: What if a publication has "transport" with type "webrtc" and not-null id?
Is "webrtc"'s TransportOptions' id always null?

File: doc/Client-Portal Protocol.md [hunk 7]
```diff
@@ -542,24 +556,124 @@ This a format for client reconnects.
...
+      }
+
+    object(P2PQuicParametersMessage)::
+      {
+       type: "quic-p2p-parameters",
+       clientTransportParameters: object(P2PQuicClientParametersMessage) | undefined,
+       serverTransportParameters: object(P2PQuicServerParametersMessage) | undefined
```
Comment (starwarfan) [inline doc/Client-Portal Protocol.md]: A "P2PQuicParametersMessage" has either "clientTransportParameters" or "serverTransportParameters"?

File: doc/Client-Portal Protocol.md (hunk 4 above)
Comment (starwarfan) [reply doc/Client-Portal Protocol.md]: I think the "soac" in object(SessionProgress") can be deleted since you moved it to "TransportProgress"

File: doc/Client-Portal Protocol.md [hunk 8]
```diff
@@ -542,24 +556,124 @@ This a format for client reconnects.
...
+    }
 
   object(SessionProgress)::
     {
      id: string(SessionId), /* StreamId returned in publishing or SubscriptionId returned in subscribing*/
      status: "soac" | "ready" | "error",
-     data: object(OfferAnswer) | object(CandidateMessage)/*If status equals “soac”*/
-          | (undefined/*If status equals “ready” and session is NOT for recording*/
-             | object(RecorderInfo)/*If status equals “ready” and session is for recording*/ )
-          | string(Reason)/*If status equals “error”*/
```
Comment (starwarfan) [inline doc/Client-Portal Protocol.md]: I think the "recorder" api has been removed from client-portal signaling. So the whole "SessionProgress" object can probably be removed since your "TransportProgress" replace it.

File: doc/Client-Portal Protocol.md (hunk 6 above)
Comment (jianjunz) [reply doc/Client-Portal Protocol.md]: It could be non-null. That's my though about reusing `PeerConnection` for multiple streams. Actually I haven't figured out what does `transportId` mean for WebRTC connections. Is it a PeerConnection ID? or transceiver ID? For now, `PeerConnection` doesn't have an ID, and transceiver cannot be shared by two tracks in the same direction. But it's quite clear for QUIC connections.

File: doc/Client-Portal Protocol.md (hunk 4 above)
Comment (jianjunz) [reply doc/Client-Portal Protocol.md]: If `RecorderInfo` is removed, it's good to remove `soac` from `SessionProgress`.

File: doc/Client-Portal Protocol.md (hunk 2 above)
Comment (jianjunz) [reply doc/Client-Portal Protocol.md]: From protocol perspective, a WebRTC publication could have `data:true` if data channel is supported, but it's not implemented right now. I'll add more description to it.

File: doc/Client-Portal Protocol.md (hunk 7 above)
Comment (jianjunz) [reply doc/Client-Portal Protocol.md]: Yes, do you think we should combine them together? Depends on browser implementations, these structures are subject to change.

File: doc/Client-Portal Protocol.md (hunk 8 above)
Comment (jianjunz) [reply doc/Client-Portal Protocol.md]: Okay, I'll remove `soac` from `SessionProgress`.

File: doc/Client-Portal Protocol.md [hunk 9]
```diff
@@ -542,24 +560,117 @@ This a format for client reconnects.
...
+      }
 **ResponseData**: undefined if **ResponseStatus** is “ok”.
 ### 3.3.15 Participant Receives Session Progress
 **NotificationName**: “progress”<br>
 
-**NotificationData**: The SessionProgress object with following definition.
+**NotificationData**: The SessionProgress or TransportProgress object with following definition.
 
-  object(SessionProgress)::
+  object(TransportProgress)::
```
Comment (starwarfan) [inline doc/Client-Portal Protocol.md]: What's the "transportId" for a "webrtc" connection since "TransportId" setting is meaning less for webrtc, "SessionId"?
Then how we distinguish "TransportProgress" and "SessionProgress", cause they are all sent through the notification name "progress"? Are they one thing?
And I have another question, for "quic" connection, "TransportId" and "SessionId" are different? Then for "quic", what's the difference between a "Transport ready status" and a "Session ready status"?

File: doc/Client-Portal Protocol.md (hunk 7 above)
Comment (starwarfan) [reply doc/Client-Portal Protocol.md]: I think we can have different types like "quic-p2p-client-parameters" and "quic-p2p-server-parameters", so readers know only one role can be used.

File: doc/Client-Portal Protocol.md [hunk 10]
```diff
@@ -514,14 +532,14 @@ This a format for client reconnects.
...
-     signaling: object(OfferAnswer) | object(CandidateMessage) | object(RemovedCandidatesMessage)
+     id: string(transportId), /* Transport ID returned in publishing or subscribing*/
+     signaling: object(OfferAnswer) | object(CandidateMessage) | object(RemovedCandidatesMessage) | object(P2PQuicParametersMessage)
     }
 
     object(OfferAnswer)::
       {
        type: "offer" | "answer",
-       sdp: string(SDP)
+       sdp: string(SDP) | null,  // WebRTC connection
```
Comment (starwarfan) [inline doc/Client-Portal Protocol.md]: null for "quic"?

File: doc/Client-Portal Protocol.md (hunk 6 above)
Comment (starwarfan) [reply doc/Client-Portal Protocol.md]: May give PeerConnection an ID later

Comment (starwarfan) [reply doc/Client-Portal Protocol.md]: If we don't support setting transportId for "webrtc" type, add a comment so people won't try.

Comment (jianjunz) [reply doc/Client-Portal Protocol.md]: Done.

File: doc/Client-Portal Protocol.md (hunk 9 above)
Comment (jianjunz) [reply doc/Client-Portal Protocol.md]: Adding a new property `type` to `TransportProgress` and `SessionProgress`? Transport ready means it's able to send or receive data. Session ready means data source or sink is ready? That's my thought about conference server. If session ready is not needed, it's okay for QUIC connections.

File: doc/Client-Portal Protocol.md (hunk 10 above)
Comment (jianjunz) [reply doc/Client-Portal Protocol.md]: Sounds like it will not be `null` because `P2PQuicParametersMessage` is added.

**Comment by jianjunz** [comment]
I think P2P QUIC will eventually be replaced by WebTransport since conference mode is C/S architecture. Implementing P2P QUIC at this time helps us smoothly bring in QUIC support and do some early experiments to optimize its performance. I hope the P2P QUIC module and protocol changes could lowly coupled with other modules.

**Comment by jianjunz** [comment]
Since QuicTransport of WebTransport has partially been implemented in Chrome, we'll skip P2P mode and move forward to WebTransport.

**Review by taste1981 — DISMISSED** [review]

**Review by taste1981 — APPROVED** [review]
Still need single PC changes to be merged to form the entire v1.2 protocol.
